### PR TITLE
[SR-3871] Unwrap multiple layers of Any/AnyObject when casting to protocol type

### PR DESCRIPTION
**Pull request by tbkka**
This fixes an oversight in the runtime dynamic cast operator:  When casting to a protocol-constrained existential type, it would unwrap at most one layer of existential containers.

As a result, code that mixed Any with AnyObject and/or Obj-C passing would be unable to cast the result to a protocol type.

Resolves SR-3871 (and some but likely not all of the duplicates)

### CODE REVIEW

**Comment by tbkka** [comment]
@swift-ci Please test Linux

**Comment by tbkka** [comment]
@swift-ci Please test macOS

**Comment by tbkka** [comment]
@swift-ci Please test Linux

**Comment by tbkka** [comment]
@swift-ci Please test Linux

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

File: stdlib/public/runtime/Casting.cpp [hunk 1]
```diff
@@ -1879,11 +1879,34 @@ static bool tryDynamicCastBoxedSwiftValue(OpaqueValue *dest,
...
+      if (isAnyObjectExistentialType(existentialType)) {
+        goto validated;
+      }
+      return false;
+    }
+    case ExistentialTypeRepresentation::Opaque: {
+      // If it's an opaque existential, unwrap it and check again
+      auto opaqueContainer = reinterpret_cast<OpaqueExistentialContainer*>(src);
+      srcType = opaqueContainer->Type;
+      src = existentialType->projectValue(src);
```
Comment (slavapestov) [inline stdlib/public/runtime/Casting.cpp]: If `flags & DynamicCastFlags::TakeOnSuccess` is true, do we destroy the innermost `src` or the original `src`? It appears to be doing the former, which I suspect is incorrect.

**Comment by tbkka** [comment]
@swift-ci Please test Linux

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

**Pull request merged**